### DISCUSSION TITLE
Simulating hundreds of channels in L+C+S+.. band

### DISCUSSION THREAD

**Discussion by devigili** [Telecominfraproject/oopt-gnpy#401] [Q&A] [answered]
Dear contributors,
I am facing an issue while simulating more than 200 channels on a huge bandwidth. I have tried to reconfigure the amplifiers but I still cannot do it. How can I avoid this Power Reduction without reducing the number of channels? Which parameters should I change?
Thanks for your attention.
![Power_Reduction_many_channels](https://user-images.githubusercontent.com/84540669/119695133-2e648c00-be4e-11eb-9c8a-b7297dbca2d4.png)

**Comment by ojnas** [comment, marked as answer]
How and where do you try to reconfigure the amplifiers? The warning you get comes from auto-design, where GNPy automatically selects an EDFA type, which means you did not specify any EDFA type in the topology file. It says that there is no EDFA type_variety in the equipment file that is "allowed_for_design" and has high enough "gain_flatmax" and/or "p_max".

Note, however, that auto-design is not aware of any Raman gain so if you have a RamanFiber with pumps followed by an EDFA you may get false warnings (or even wrong results) as discussed in https://github.com/Telecominfraproject/oopt-gnpy/issues/398. In this case you should really specify the EDFA type_variety explicitly.

**Reply by devigili** [reply]
I was reconfiguring the amplifiers in topology.json, basically by increasing their gain_flatmax" and "p_max" 
I found out in the topology_auto_design that some amplifiers `"Edfa0"` were there autogenerated. I am still trying to adjust the `"gain_flatmax"` ,"`p_max"` values enough in the equipment and the "gain_target" in the topology to avoid these power reduction.

I am using  `RamanFibers` but without pumps, in order to realize simulation adopting the GGN model.

**Reply by ojnas** [reply]
>I was reconfiguring the amplifiers in topology.json, basically by increasing their gain_flatmax" and "p_max"

The "gain_flatmax" and "p_max" should be defined in the equipment file, not in the topology file.

>I found out in the topology_auto_design that some amplifiers "Edfa0" were there autogenerated.

Yes, this is also a feature of auto-design. You can tell auto-design to not insert an amplifier after a fiber or ROADM by inserting a Fused element, see some info here (I think we should probably make it easier to disable the automatic EDFA creation globally but for now, the Fused element is what you can use):

https://gnpy.readthedocs.io/en/master/json.html#global-parameters

**Comment by jktjkt** [comment]
Apart from what @ojnas said, I'm also interested in the actual optical bandwidth that you're propagating. You'll probably get reasonably accurate numbers if you limit yourself to propagation within the C-band, or perhaps a substantial chunk of the C-band. But once you hit anything wider, all bets are off. You will still get *some* numbers out of **the stable version of** GNPy, but these might as well be just numbers that do not reflect reality.

**Reply by ojnas** [reply]
> But once you hit anything wider, all bets are off. You will still get _some_ numbers out of GNPy, but these might as well be just numbers that do not reflect reality.

While it's good to be aware of possible limitations, I think this is a bit overly pessimistic. I think it's fully reasonable to try out GNPy for C+L-band propagation, at least if you take Raman effects into account.

**Reply by devigili** [reply]
Once I will be able to transmit more channels I will try to develope a new model for the amplifiers in order to get more realistic numbers.

**Reply by jktjkt** [reply]
I should have emphasized that what I wrote above applies to the past release of GNPy (based on what I remember from the Polito team). There's a patch series in the works which improves this a lot. Perhaps @AndreaDAmico or others from the POLITO team could chime in?

**Reply by AndreaDAmico** [reply]
Virtually, the current version of GNPy can simulate the C+L scenario, but the accuracy of the result is not guaranteed. In order to have a reasonable simulation, the following issues have to be addressed:
- The RamanFiber can already simulate the propagation of a C+L spectrum. There would be a computational time increasing due to the high number of channels; in the new patch series the Raman solver scales better with the number of channels. To obtain an accurate result on such a wide spectrum, a loss coefficient profile along the frequencies have to be provided;
- In  realistic scenarios, the C+L spectrum is amplified separately by two distinct devices, the EDFA in C and L band, respectively. To date, GNPy does not include this option, but I think an adequate configuration of the current EDFA implementation can be provided depending on the required accuracy. The main issue is the definition of the noise figure and the tilt as an accurate simulation have to include two different noise figure profiles for the two EDFA types and, having a non negligible Raman effect, a proper tilt profile have to be provided.

**Reply by devigili** [reply]
Dear @AndreaDAmico,
Could you please give me a hint about how to assign a specific noise figure and tilt per each band? I believe that there are two way:
1. Create in the equipment some new Edfa `"type_variety":`  with band-dependent NF. Then modify the core code in order to be able to add more than one `"type": "Edfa"`, with different "type_variety": per each node in the topology.json. Which pyton file should I manipulate?

2. use the `"type": "Edfa"`, with `"type_def": "advanced_model", ` and configure the std_medium_gain_advanced_config.json file accordingly.  At my understanding `"type_def": "advanced_model",` relies on the third order polynomial $𝑔 (𝑦) = a𝑦^3 + b𝑦^2+ c𝑦 + d$ , $NF = 𝑔 (𝐻max − 𝐻)$. What is the relation between a,b,c,d and `"gain_ripple", "dgt", "nf_ripple" "nf_fit_coeff":`

Which one of the paths would you consider more correct? 
Thank you very much

**Reply by ojnas** [reply]
About 2:
It's `nf_fit_coeff` that holds the polynomial coefficients but I don't see how they help modeling separate NF, gain and tilt for different bands:
https://github.com/Telecominfraproject/oopt-gnpy/blob/8c0eac1bdc1e95effbb4088544dee5fc66a31273/gnpy/core/elements.py#L695

About 1:
Something like that sounds like the right approach. You would need to modify at least both the input file parsing code in gnpy/tools/json_io.py and the Edfa propagation code in gnpy/core/elements.py. The Edfa propagation code would need to split up the SpectralInformation in separate bands, propagate each band independently, and then combine the bands again.

**Reply by AndreaDAmico** [reply]
I agree, the first approach seems to be the right choice for your use case. You have to define a new 'type_variety', but I think there is no need to split the spectrum, you can use the Edfa attribute `interpol_nf_ripple` which is calculated using the parameters `f_min`, `f_max` and `nf_ripple`, the latter fixed as a step function (even better a precise profile along the frequency if you know it).  The same is true for the tilt, defining the rigth `dgt`, you should be able to obtain the correct profile.

**Reply by ojnas** [reply]
@AndreaDAmico, maybe your proposal works for the NF ripple and DGT but what if the C- and L-band EDFAs have different NF vs gain characteristics and needs different parameters (polynomial coefficients or nf_min/nf_max for example)?

**Reply by AndreaDAmico** [reply]
@ojnas, it would depend on the available NF characterization, but in this case I would rather define a new 'type_def' and include in the method `_calc_nf` an expression to evaluate the proper NF from those parameters. The main issue that I see in dividing the spectrum is the management of the pref and the feedback control that, in my opinion, is not trivial. Indeed, what I am suggesting here works for a temporal solution specific for the case under discussion, for a future contribution that would enable the C+L scenario to be simulated in the stable version of GNPy, I agree that splitting the spetcrum is the correct solution, but I would change the Edfa implementation more in depth.

**Comment by Pedro-Gonzalez-Gil** [comment]
Hi guys. Sorry to get into this discussion but I found the C+L topic
quite interesting :). If I may, two quick questions about GNPy below

   - Is GNPy able to simulate the power transfer from the C band to the L
   band?
   - Is there any option to simulate the use of ASE light sources
   (idlers/loads)?

Thanks!
Pedro

**Reply by jktjkt** [reply]
There's no way to inject an ASE-like (unmodulated) signal into GNPy today, and I think that the current GNPy assumes that all signals come from "a modern transponder". But my background is SW engineering, not optics, so I cannot say much besides that.

However, if you're looking for "worst case scenarios" with a full channel load, I think GNPy actually gives the conservative answer. I.e., if you simulate for, say, 70 channels and only provision 10 of them in your network, the network should not perform worse than the simulation results predict. And when it comes to the GSNR estimates in GNPy, they are given at the "full channel load" as defined via [`f_min`, `f_max` and `spacing` in the `SpectralInformation`](https://gnpy.readthedocs.io/en/master/json.html#spectralinformation).

**Reply by jktjkt** [reply]
Also, may I suggest to [start a new question thread](https://github.com/Telecominfraproject/oopt-gnpy/discussions/new) for additional questions? That way it's easier for everybody to follow the discussion. It is currently not possible to do that via replying to an e-mail, I'm afraid, so you'll have to go to Github for this. Thanks!

**Reply by ojnas** [reply]
Hi @Pedro-Gonzalez-Gil,

> Is GNPy able to simulate the power transfer from the C band to the L band? 

Yes, it already does this by default if you use the RamanFiber element. As @jktjkt mentioned in the other thread, there is work in progress to improve the Raman simulation code, e.g. to speed up the computation time, which will help. In addition, there is work in progress to support propagation of mixed baudrates / flexible grid and non-contiguous spectrum, which I guess will also help modelling a realistic C+L-band system.

> Is there any option to simulate the use of ASE light sources (idlers/loads)?

I don't have any experience with C+L-band systems but I thought that "idlers" (or "dummy light") was something mainly used in subsea systems with constant power amplifiers and not directly related to C+L-band. Anyway, I agree with @jktjkt that simulating a system with full channel load should result in worst case performance, i.e. replacing some real channels with ASE noise loading should not result in worse performance. On the other hand, I think that GNPy should in principle also be able to handle propagation of "white noise channels" since the nonlinear fiber model does not really care about the modulation (@AndreaDAmico, please correct me if I'm wrong). However, this would need some thought about how the input to the simulation should be defined.

EDIT: I see that @AndreaDAmico replied almost simultaneously in the other thread so check his answer also.

**Reply by AndreaDAmico** [reply]
Just to add a bit of information, GNPy use a modified version of the Gaussian model (GN) in order to calculate the NLI generated along the fiber. Namely, the channels are considered Gaussian modulated. This represents an overestimation of the NLI generation if the channels are actually modulated with a real modulation formats, but this is not the case if they are "ASE noise loaded" channels. 
In any case, this difference is not significant in most realistic scenarios and, as @ojnas pointed out, GNPy can handle the propagation of  "white noise channels", which I presume are shaped as a standard channel, but generated by an ASE source. 
On the other hand, if the white noise is not shaped and occupies a wide bandwidth, it is necessary to separate it in multiple channels in order to obtain the correct simulation of both the Raman scattering effect and the NLI generation.  This separation can be done as @jktjkt described; the best solution would be setting the baud_rate (in Hz units) equal to the spacing. The number of channels that are needed in this case is a trade-off between the accuracy (high number of channels) and computational time (low number of channels).
